esp-bootloader-esp-idf/src/partitions.rs

```diff
@@ -0,0 +1,757 @@
+        self.flags() & 0b10 != 0
+    }
+
+    /// The partition type (type and sub-type).
+    pub fn partition_type(&self) -> PartitionType {
+        match self.raw_type() {
+            0 => PartitionType::App(self.raw_subtype().try_into().unwrap()),
+            1 => PartitionType::Data(self.raw_subtype().try_into().unwrap()),
+            2 => PartitionType::Bootloader(self.raw_subtype().try_into().unwrap()),
+            3 => PartitionType::PartitionTable(self.raw_subtype().try_into().unwrap()),
+            _ => unreachable!(),
+        }
+    }
+
+    /// Provides a "view" into the partition allowing to read/write the
+    /// partition contents by using the given [embedded_storage::Storage] and/or
+    /// [embedded_storage::ReadStorage] implementation.
+    pub fn as_embedded_storage<F>(&'a self, flash: &'a mut F) -> FlashRegion<'a, F>
+    where
+        F: embedded_storage::ReadStorage,
+    {
+        FlashRegion { raw: self, flash }
+    }
+}
+
+impl core::fmt::Debug for PartitionEntry<'_> {
+    fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
+        f.debug_struct("PartitionEntry")
+            .field("magic", &self.magic())
+            .field("raw_type", &self.raw_type())
+            .field("raw_subtype", &self.raw_subtype())
+            .field("offset", &self.offset())
+            .field("len", &self.len())
+            .field("label", &self.label_as_str())
+            .field("flags", &self.flags())
+            .field("is_read_only", &self.is_read_only())
+            .field("is_encrypted", &self.is_encrypted())
+            .finish()
+    }
+}
+
+#[cfg(feature = "defmt")]
+impl defmt::Format for PartitionEntry<'_> {
+    fn format(&self, fmt: defmt::Formatter) {
+        defmt::write!(
+            fmt,
+            "PartitionEntry (\
+            magic = {}, \
+            raw_type = {}, \
+            raw_subtype = {}, \
+            offset = {}, \
+            len = {}, \
+            label = {}, \
+            flags = {}, \
+            is_read_only = {}, \
+            is_encrypted = {}\
+            )",
+            self.magic(),
+            self.raw_type(),
+            self.raw_subtype(),
+            self.offset(),
+            self.len(),
+            self.label_as_str(),
+            self.flags(),
+            self.is_read_only(),
+            self.is_encrypted()
+        )
+    }
+}
+
+/// Errors which can be returned.
+#[derive(Debug, PartialEq, Eq, Clone, Copy, Hash, strum::Display)]
+#[cfg_attr(feature = "defmt", derive(defmt::Format))]
+pub enum Error {
+    /// The partition table is invalid.
+    Invalid,
+    /// An operation tries to access data that is out of bounds.
+    OutOfBounds,
+    /// An error which originates from the embedded-storage implementation.
+    StorageError,
+    /// The partition is write protected.
+    WriteProtected,
+}
+
+impl core::error::Error for Error {}
+
+/// A partition table.
+pub struct PartitionTable<'a> {
+    binary: &'a [u8],
```

Comment:
This might be a harder sell :sweat_smile: but I think it works well with my earlier suggestion,
```suggestion
    binary: &'a [[u8; RAW_ENTRY_LEN]],
```